---
title: 'phg-1'
date: 2025-06-02
permalink: /posts/2025/06/phg-1/
tags:
  - chapter1
  - phg
---

# 1.Build Apptainer
Connect to CSC.puhti

`ssh chenghao@puhti.csc.fi`
## 1.1 Download data

### 1.1.1 Install NCBI datasets package

`apptainer build datasets.sif docker://staphb/ncbi-datasets`

### 1.1.2 Prepare accessions.txt
`nano accessions.txt`
```
GCA_902167145.1
GCA_910593975.1
GCA_902167075.1
GCA_902167055.1
GCA_902166975.1
GCA_003709335.1
GCA_001984235.2
GCA_001990705.1
GCA_902167065.1
GCA_902167115.1
GCA_022117705.1
GCA_902167035.1
GCA_003704525.1
GCA_002237485.1
GCA_019096015.1
GCA_019095995.1
GCA_001644905.2
GCA_963692765.1
GCA_963693405.1
GCA_963693425.1
```

### 1.1.3 Prepare Slurm script

`nano download_genomes.slurm`
```
#!/bin/bash
#SBATCH --account=project_2013932   
#SBATCH --partition=small               
#SBATCH --time=04:00:00
#SBATCH --mem=16G
#SBATCH --job-name=genome_download
#SBATCH --output=download.out
#SBATCH --error=download.err

# OUT_DIR=/scratch/project_2013932/phg/data

# mkdir -p $OUT_DIR

apptainer exec datasets.sif datasets download genome accession \
  --inputfile /scratch/project_2013932/phg/accessions.txt \
  --include genome \
#  --dehydrated \
  --no-progressbar \
  --filename genomes.zip
#  --output-folder $OUT_DIR


# apptainer exec datasets.sif datasets rehydrate --directory $OUT_DIR
```
Submitting the batch job

`sbatch download_genomes.slurm`

`unzip genomes.zip -d genomes`

`find data -name "*.fna*" -exec cp {} data/ \;`


## 1.2 Install PHG
### 1.2.1 Creat a def file for apptainer
`nano phg2.def`

phg2.def was used to built the environment for PHG
```
BootStrap: docker
From: debian:bullseye-slim

%labels
    Author: YourName
    Version: PHG2-release-2.4.61.216-libmamba-agc3.1
    Description: PHG v2.4 container with Java 17, Miniconda, libmamba solver, AGC v3.1

%post

    export TMPDIR=/scratch/project_2013932/phg/chenghao_conda_cache/tmp_build_dir
    mkdir -p $TMPDIR
    export DEBIAN_FRONTEND=noninteractive

    # 安装系统依赖
    apt-get update && apt-get install -y --no-install-recommends \
        wget curl git unzip bzip2 ca-certificates \
        build-essential zlib1g-dev libbz2-dev liblzma-dev \
        libcurl4-openssl-dev libssl-dev \
        openjdk-17-jdk samtools tabix cmake autoconf automake \
        && apt-get clean && rm -rf /var/lib/apt/lists/*

    # 安装 Minimap2
    export INSTALL_DIR=/opt
    mkdir -p $INSTALL_DIR
    cd $INSTALL_DIR && git clone https://github.com/lh3/minimap2.git && cd minimap2 && make -j$(nproc)
    ln -s $INSTALL_DIR/minimap2/minimap2 /usr/local/bin/minimap2

    # 安装 AnchorWave
    cd $INSTALL_DIR && git clone https://github.com/baoxingsong/anchorwave.git && cd anchorwave
    mkdir build && cd build
    cmake ..
    make -j$(nproc)
    ln -s $INSTALL_DIR/anchorwave/build/anchorwave /usr/local/bin/anchorwave

    # 安装 PHG release tar 包
    cd $INSTALL_DIR
    curl -LO https://github.com/maize-genetics/phg_v2/releases/download/2.4.61.216/PHGv2-v2.4.tar
    mkdir phg-v2.4 && tar -xvf PHGv2-v2.4.tar -C phg-v2.4
    PHG_EXE=$(find $INSTALL_DIR/phg-v2.4 -type f -name phg | head -n 1)
    chmod +x "$PHG_EXE"
    ln -s "$PHG_EXE" /usr/local/bin/phg

    # 安装 Miniconda
    MINICONDA_DIR=/opt/miniconda
    wget --quiet https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O /tmp/miniconda.sh
    bash /tmp/miniconda.sh -b -p $MINICONDA_DIR
    rm /tmp/miniconda.sh
    export PATH=$MINICONDA_DIR/bin:$PATH

    # 设置 Conda 缓存路径为 /scratch
    mkdir -p /scratch/project_2013932/phg/chenghao_conda_cache/pkgs \
             /scratch/project_2013932/phg/chenghao_conda_cache/cache \
             /scratch/project_2013932/phg/chenghao_conda_cache/envs \
             /scratch/project_2013932/phg/chenghao_conda_cache/tmp

    export CONDA_PKGS_DIRS=/scratch/project_2013932/phg/chenghao_conda_cache/pkgs
    export CONDA_CACHE_DIR=/scratch/project_2013932/phg/chenghao_conda_cache/cache
    export CONDA_ENVS_DIRS=/scratch/project_2013932/phg/chenghao_conda_cache/envs
    export TMPDIR=/scratch/project_2013932/phg/chenghao_conda_cache/tmp

    . $MINICONDA_DIR/etc/profile.d/conda.sh

    # 安装 libmamba solver
    conda update -n base -c defaults conda -y
    conda install -n base -c conda-forge conda-libmamba-solver libmamba libmambapy -y
    conda config --set solver libmamba
    echo "solver: libmamba" > /opt/.condarc
    export CONDARC=/opt/.condarc

    # 安装 AGC v3.1
    mkdir -p $INSTALL_DIR/agc-bin && cd $INSTALL_DIR/agc-bin
    wget https://github.com/refresh-bio/agc/releases/download/v3.1/agc-3.1_x64_linux.tar.gz
    tar -xzf agc-3.1_x64_linux.tar.gz
    chmod +x agc-3.1_x64_linux/agc
    ln -s $INSTALL_DIR/agc-bin/agc-3.1_x64_linux/agc /usr/local/bin/agc

%environment
    export PATH=/usr/local/bin:/opt/minimap2:/opt/anchorwave/build:/opt/miniconda/bin:$PATH
    export JAVA_HOME=/usr/lib/jvm/java-17-openjdk-amd64

    export CONDA_PKGS_DIRS=/scratch/project_2013932/phg/chenghao_conda_cache/pkgs
    export CONDA_CACHE_DIR=/scratch/project_2013932/phg/chenghao_conda_cache/cache
    export CONDA_ENVS_DIRS=/scratch/project_2013932/phg/chenghao_conda_cache/envs
    export TMPDIR=/scratch/project_2013932/phg/chenghao_conda_cache/tmp
    export CONDARC=/opt/.condarc

%runscript
    echo "Welcome to PHG2 container (v2.4, Java 17, libmamba solver, AGC v3.1)."
    exec "$@"
```
Build apptainer and test
```
apptainer build phg2.sif phg2.def
apptainer exec phg2.sif phg --help

apptainer shell phg2.sif
phg
minimap2 --version
anchorwave --version
java -version
```

### 1.2.2 Install conda environment
Initial conda

`
source /opt/miniconda/etc/profile.d/conda.sh
`
`
source /opt/conda/etc/profile.d/conda.sh
`
`
export PATH=/scratch/project_2013932/phg/bin:$PATH
`

Install conda of PHG

`phg setup-environment`

`conda activate phgv2-conda`

# 2. Initialize TileDB instances
```
phg initdb \
    --db-path vcf_dbs \
    --gvcf-anchor-gap 1000000 \
    --hvcf-anchor-gap 1000 \
    --conda-env-prefix /scratch/project_2013932/phg/chenghao_conda_cache/envs/phgv2-tiledb
```

# 3. Prepare Assembly FASTA files
prepare keyfile
```
/data/GCA_902167145.1_Zm-B73-REFERENCE-NAM-5.0_genomic.fna	Ref
/data/GCA_910593975.1_Zm-B104-REFERENCE-CORTEVA-1.0_genomic.fna	B104
/data/GCA_902167075.1_Zm-B97-REFERENCE-NAM-1.0_genomic.fna	B97
/data/GCA_902167055.1_Zm-CML103-REFERENCE-NAM-1.0_genomic.fna	CML103
/data/GCA_902166975.1_Zm-CML247-REFERENCE-NAM-1.0_genomic.fna	CML247
/data/GCA_003709335.1_Zm-DK105-REFERENCE-TUM-1.0_genomic.fna	DK105
/data/GCA_001984235.2_Zm-EP1-REFERENCE-TUM-1.0_genomic.fna	EP1
/data/GCA_001990705.1_Zm-F7-REFERENCE-TUM-1.0_genomic.fna	F7
/data/GCA_902167065.1_Zm-Il14H-REFERENCE-NAM-1.0_genomic.fna	IL14H
/data/GCA_902167115.1_Zm-Ki11-REFERENCE-NAM-1.0_genomic.fna	Ki11
/data/GCA_022117705.1_Zm-Mo17-REFERENCE-CAU-T2T-assembly_genomic.fna	MO17
/data/GCA_902167035.1_Zm-NC358-REFERENCE-NAM-1.0_genomic.fna	NC358
/data/GCA_003704525.1_Zm-PE0075-REFERENCE-TUM-1.0_genomic.fna	PE0075
/data/GCA_002237485.1_Zm-PH207-REFERENCE_NS-UIUC_UMN-1.1_genomic.fna	PH207
/data/GCA_019096015.1_Zm_PHB47_v1_genomic.fna	PHB47
/data/GCA_019095995.1_Zm_PHJ40_v1_genomic.fna	PHJ40
/data/GCA_001644905.2_Zm-W22-REFERENCE-NRGENE-2.0_genomic.fna	W22
/data/GCA_963692765.1_Zm-Mo44-Draft-G2F-1.0_genomic.fna	Mo44
/data/GCA_963693405.1_Zm-MoG-Draft-G2F-1.0_genomic.fna	MoG
/data/GCA_963693425.1_Zm-PHW65-Draft-G2F-1.0_genomic.fna	PHW65
```

```
phg prepare-assemblies \
    --keyfile data/annotation_keyfile.txt \
    --threads 10 \
    --output-dir output/updated_assemblies
```
# 4. Compress FASTA files
```
phg agc-compress \
    --db-path vcf_dbs \
    --fasta-list data/assemblies_list.txt \
    --reference-file output/updated_assemblies/Ref.fa \
    --conda-env-prefix /scratch/project_2013932/phg/chenghao_conda_cache/envs/phgv2-tiledb
```

# 5. Create reference ranges
```
phg create-ranges \
    --gff data/B73v5.gff3 \
    --reference-file output/updated_assemblies/Ref.fa \
    --boundary gene \
    --pad 500 \
    --range-min-size 500 \
    -o output/ref_ranges.bed
```

# 6. Align assemblies
`nano phg_align_assemblies.slurm`
```
#!/bin/bash
#SBATCH --account=project_2013932
#SBATCH --time=24:00:00                # 每个任务最多运行 16 小时
#SBATCH --nodes=1
#SBATCH --ntasks=20            # 每个任务使用 4 个核心
#SBATCH --mem-per-cpu=12G                      # 每个任务分配 64GB 内存
#SBATCH --partition=small
#SBATCH --job-name="phg_align_array"
#SBATCH --mail-type=END
#SBATCH --array=0,2,3,5,6,7,9,11,16,19                   # 假设你有 20 个样本

export _JAVA_OPTIONS="-Xmx180G"

# ==== Apptainer 容器路径 ====
CONTAINER=/scratch/project_2013932/phg/phg2.sif

# ==== 输入的命令文件，每一行是 phg align-assemblies 命令 ====
COMMAND_FILE=/scratch/project_2013932/phg/output/slurm_align_file.txt

# ==== 绑定目录（确保这些路径在容器中可见） ====
apptainer exec \
    --bind /scratch/project_2013932/phg:/scratch/project_2013932/phg \
    $CONTAINER \
    bash -c "
        source /opt/miniconda/etc/profile.d/conda.sh
        conda activate /scratch/project_2013932/phg/chenghao_conda_cache/envs/phgv2-conda

        INPUTFILE=${COMMAND_FILE}
        IFS=\$'\n' read -d '' -r -a LINES < \$INPUTFILE
        LINE=\${LINES[\$SLURM_ARRAY_TASK_ID]}
        echo Running command: \$LINE
        eval \$LINE
        RET=\$?
        if [ \$RET -eq 0 ]; then
            echo -e \"\$(date +\"%D  %r\")\tSuccess: \$LINE\"
            exit 0
        else
            echo -e \"\$(date +\"%D  %r\")\tFailed: \$LINE\"
            exit 1
        fi
    "
```
```
sbatch phg_align_assemblies.slurm
```